addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,448 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	clusterapiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	clusterapiutil "sigs.k8s.io/cluster-api/util"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	kapppkgiv1alpha1 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/packaging/v1alpha1"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/types"
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	"github.com/vmware-tanzu/tanzu-framework/addons/predicates"
+	runtanzuv1alpha3 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
+	tkrconstants "github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkr/pkg/constants"
+)
+
+const kappCtrlPkgPrefix = "kapp-controller"
+
+type ClusterRole int
+
+const (
+	clusterRoleManagement ClusterRole = iota
+	clusterRoleWorkload
+)
+
+// PackageInstallStatusReconciler contains the reconciler information for PackageInstallStatus controller
+type PackageInstallStatusReconciler struct {
+	Ctx        context.Context
+	Log        logr.Logger
+	Client     client.Client
+	Scheme     *runtime.Scheme
+	controller controller.Controller
+	Tracker    *remote.ClusterCacheTracker
+	// aggregatedAPIResourcesClient is used when it is required to directly read from the server and not to use object caches.
+	aggregatedAPIResourcesClient client.Client
+}
+
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=packageinstallstatuses,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=addons.tanzu.vmware.com,resources=packageinstallstatuses/status,verbs=get;update;patch
+
+// Reconcile performs the reconciliation action for the controller; which is reflecting the reconciliation status of each core/additional package
+// into corresponding ClusterBootstrap resource of the cluster
+func (r *PackageInstallStatusReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.Log.WithValues(constants.ClusterNamespaceLogKey, req.Namespace, constants.ClusterNameLogKey, req.Name)
```

Comment:
Just a suggestion. Let's do a follow-up on logging. A while back in CAPI we moved away from per-reconciler loggers to per-request loggers (by reading them out of the context)

In my opinion it has a few advantages, would profit from latest improvements in controller-runtime and aligns to the upstream contextual logging KEP.

We are currently working on a logging guideline for CAPI and I think a lot of it would/could apply here too.

But of course absolutely up to you and the topic is independent of this PR.

Reply:
P.S. Similar topic with the context itself.

Reply:
/cc @vijaykatam

Reply:
Made this change

Reply:
I'm not sure if the change should be made in this PR. I think it would be better to keep this controller consistent with the other ones and then consider a refactoring for all controllers.

What I wanted to hint add (but I would provide more details) is that:
* We should use the context passed into the Reconcile func and propagate it through all functions
    * This makes it future proof for things like passing around timeouts, canceling ongoing reconciles during shutdown, future tracing implementation in controller-runtime, probably more
    * In general if available in my opinion it's a good practice to use per-request contexts rather than ones created at startup of a binary
* We should get the logger from the context passed into the Reconcile func
    * controller-runtime already adds a few useful k/v pairs to that logger

But as all other controllers are using a "static" logger and context in reconcile I think this is something which should ideally be discussed as follow-up. I can also provide more details then and show how it all fits together in CAPI today. And then it can be decided if you/tanzu-framework wants to align to the pattern we use in CAPI>

Reply:
Thanks for the comment Stefan. Good idea that better to keep it consistent. I'll revert my changes around logging for now and we'll definitely get your advice to hear about best practices in CAPI on this, some time in the near future

Reply:
@vijaykatam Filed this issue https://github.com/vmware-tanzu/tanzu-framework/issues/1948